### DISCUSSION TITLE
RFC about a set of contributions to solved some issues pending

### DISCUSSION THREAD

**Discussion by llaville** [composer-unused/composer-unused#630] [Show and tell]
Hello everybody,

I would like here to propose a version of `composer-unused/composer-unused` (and `composer-unused/symbol-parser`) 
which have some little improvements against current versions [1] and [2].

[1] : https://github.com/composer-unused/composer-unused/releases/tag/0.8.11
[2] : https://github.com/composer-unused/symbol-parser/releases/tag/0.2.2

Of course these versions are not yet official, and all patches includes are not yet validated by @icanhazstring, 
but if you and me want a better application/library with less issues, any feedback/tests will be welcome !

Hope all my explains will be enough to understand.
Listening on you !

TL;DR;

I would like especially highlight some bug fixes : 

on `symbol-parser`

- [Call to undefined method `PhpParser\Node\Expr\Closure::getParts()` in `DefinedSymbolCollector`](https://github.com/composer-unused/symbol-parser/issues/122) : initial reporter @Wirone
- [`DefinedSymbolCollector` did not return traits and namespaces](https://github.com/composer-unused/symbol-parser/issues/136) : initial reporter @llaville
- [Undesired namespace consolidation](https://github.com/composer-unused/symbol-parser/issues/142) : initial reporter @ComiR
- [`ConsumedSymbolCollector` accept invalid strategies](https://github.com/composer-unused/symbol-parser/issues/148) : initial reporter @llaville

on `composer-unused`

- [Call to undefined method `PhpParser\Node\Expr\Variable::getParts()`](https://github.com/composer-unused/composer-unused/issues/610) : initial reporter @simPod 

An also highlight some new feature I've added :

on `symbol-parser`

- [Add a new `DefineStrategy` to replace the `DefinedSymbolCollector`](https://github.com/composer-unused/symbol-parser/discussions/137)
- [Add a new NodeSymbolCollector for debugging or other context](https://github.com/composer-unused/symbol-parser/discussions/145) : learn more on demo at https://github.com/composer-unused/symbol-parser/discussions/151

Now for details of this un-official version.

I've two forks of these projects (with at least a `contribs` branch that includes all patches) :

- <https://github.com/llaville/symbol-parser/tree/contribs> (latest commit `9e4bb1c` verified only by me)
- <https://github.com/llaville/composer-unused/tree/contribs> (latest commit `7737a17` verified only be me)

If you want to give it a try, and contribute yourself to improve stability of this project, it's your turn : test and report please !

**IMPORTANT** : 
I've introduces specialized version of `NameContext` and `NameResolver` components to detect elements identified by special class names `self` and `parent`.
But it's not applied on `composer-unused/composer-unused` source code of my fork.
It'll allow you to see the different, without and with these new specialized versions.

Replace all import `use PhpParser\NodeVisitor\NameResolver` by `use ComposerUnused\SymbolParser\Parser\PHP\NameResolver`.

For example :

```text
diff --git a/src/Symbol/ConsumedSymbolLoaderBuilder.php b/src/Symbol/ConsumedSymbolLoaderBuilder.php
index 8327777..32f9e0b 100644
--- a/src/Symbol/ConsumedSymbolLoaderBuilder.php
+++ b/src/Symbol/ConsumedSymbolLoaderBuilder.php
@@ -7,13 +7,14 @@ namespace ComposerUnused\ComposerUnused\Symbol;
 use ArrayIterator;
 use ComposerUnused\SymbolParser\File\FileContentProvider;
 use ComposerUnused\SymbolParser\Parser\PHP\AutoloadType;
+use ComposerUnused\SymbolParser\Parser\PHP\NameResolver;
 use ComposerUnused\SymbolParser\Parser\PHP\SymbolCollectorInterface;
 use ComposerUnused\SymbolParser\Parser\PHP\SymbolNameParser;
 use ComposerUnused\SymbolParser\Symbol\Loader\FileSymbolLoader;
 use ComposerUnused\SymbolParser\Symbol\Loader\SymbolLoaderInterface;
 use ComposerUnused\SymbolParser\Symbol\Provider\FileSymbolProvider;
 use PhpParser\Lexer\Emulative;
-use PhpParser\NodeVisitor\NameResolver;
+//use PhpParser\NodeVisitor\NameResolver;
 use PhpParser\ParserFactory;
 use SplFileInfo;

```
